src/tools/cgcreate.c

```diff
@@ -221,7 +226,19 @@ int main(int argc, char *argv[])
 		if (dirm_change | filem_change)
 			cgroup_set_permissions(cgroup, dir_mode, file_mode, tasks_mode);
 
-		ret = cgroup_create_cgroup(cgroup, 0);
+		if (create_scope) {
+			struct cgroup_systemd_scope_opts opts;
+
+			/*
+			 * Intentionally ignoring the return code because we are providing a
+			 * valid pointer
+			 */
+			(void)cgroup_set_default_scope_opts(&opts);
+
+			ret = cgroup_create_scope2(cgroup, 0, &opts);
+		} else {
+			ret = cgroup_create_cgroup(cgroup, 0);
```

Comment:
When creating a delegated slice and scope, the `<slice>.name/<scope>.scope` should be written to `/var/run/libcgroup/systemd`, otherwise the delegated slice/scope will not be appended to the default cgroup path. Consider the following example:
```
# ./src/tools/cgcreate -c -g cpu:foo.slice/bar.scope
# ./src/tools/cgset -r cpu.weight=50 /
/root/libcgroup/src/tools/.libs/cgset: cgroup modify error: Cgroup, requested group parameter does not exist
```
Some like:
```
diff --git a/include/libcgroup/systemd.h b/include/libcgroup/systemd.h
index 39f6974..9a57576 100644
--- a/include/libcgroup/systemd.h
+++ b/include/libcgroup/systemd.h
@@ -119,6 +119,8 @@ int cgroup_add_systemd_opts(const char * const conf, const char * const value);
  */
 void cgroup_cleanup_systemd_opts(void);
 
+int cgroup_write_systemd_default_cgroup(const char * const slice,
+                                       const char * const scope);
 #ifdef __cplusplus
 } /* extern "C" */
 #endif
diff --git a/src/config.c b/src/config.c
index 6b4fe06..be50420 100644
--- a/src/config.c
+++ b/src/config.c
@@ -1968,7 +1968,7 @@ static int remove_systemd_default_cgroup_file(void)
  * cgroup_set_default_systemd_cgroup() for setting
  * systemd_default_cgroup used to form the cgroup path.
  */
-static int cgroup_write_systemd_default_cgroup(const char * const slice,
+int cgroup_write_systemd_default_cgroup(const char * const slice,
                                               const char * const scope)
 {
        FILE *systemd_def_cgrp_f;
@@ -2172,4 +2172,8 @@ int cgroup_alloc_systemd_opts(const char * const config, const char * const valu
 void cgroup_cleanup_systemd_opts(void) { }
 
 void cgroup_set_default_systemd_cgroup(void) { }
+int cgroup_write_systemd_default_cgroup(const char * const slice,
+                                              const char * const scope) {
+       return 1;
+}
 #endif
diff --git a/src/libcgroup.map b/src/libcgroup.map
index 28ce7d4..b0f0356 100644
--- a/src/libcgroup.map
+++ b/src/libcgroup.map
@@ -156,4 +156,5 @@ CGROUP_3.0 {
        cgroup_get_controller_name;
        cgroup_create_scope2;
        cgroup_set_default_systemd_cgroup;
+       cgroup_write_systemd_default_cgroup;
 } CGROUP_2.0;
diff --git a/src/tools/cgcreate.c b/src/tools/cgcreate.c
index 8a46a8e..29f2417 100644
--- a/src/tools/cgcreate.c
+++ b/src/tools/cgcreate.c
@@ -228,6 +228,9 @@ int main(int argc, char *argv[])
 
                if (create_scope) {
                        struct cgroup_systemd_scope_opts opts;
+                       char slice[FILENAME_MAX];
+                       char *scope;
+                       int len;
 
                        /*
                         * Intentionally ignoring the return code because we are providing a
@@ -236,6 +239,22 @@ int main(int argc, char *argv[])
                        (void)cgroup_set_default_scope_opts(&opts);
 
                        ret = cgroup_create_scope2(cgroup, 0, &opts);
+                       if (!ret) {
+                               scope = strstr(cgroup->name, "/");
+                               len = strlen(cgroup->name) - strlen(scope);
+                               strncpy(slice, cgroup->name, len);
+                               slice[len] = '\0';
+                               scope++;
+                               ret = cgroup_write_systemd_default_cgroup(slice, scope);
+                               if (!ret) {
+                                       err("%s: failed to write default %s/%s to ",
+                                               argv[0], slice, scope);
+                                       err("/var/run/libcgroup/systemd\n");
+                                       ret = 1;
+                                       goto err;
+                               }
+                               ret = 0;
+                       }
                } else {
                        ret = cgroup_create_cgroup(cgroup, 0);
                }

```
```
# ./src/tools/cgcreate -c -g cpu:foo.slice/bar.scope
# ./src/tools/cgset -r cpu.weight=50 /
# ./src/tools/cgget -r cpu.weight /
/:
cpu.weight: 50
```

or there can be a lightweight version of `cgroup_write_systemd_default_cgroup()`, implemented in the `cgcreate` tool, instead of exporting `cgroup_write_systemd_default_cgroup()`, that would effectively do:
- write `slice/scope` to `/var/run/libcgroup/systemd`
- skip the check to check if the cgroup (slice/scope) exists, because the `cgroup_create_scope2()` successful return means, the cgroup exists.

Reply:
Good idea.  I'll play around with it.